src/ripple/ledger/View.h

```diff
@@ -418,6 +420,647 @@ transferXRP(
+        return tecINTERNAL;
+
+    if (deltaAmt == beast::zero)
+        return tesSUCCESS;
+
+    // can't lock or unlock a zero balance
+    if (balance == beast::zero)
+    {
+        JLOG(j.trace()) << "trustAdjustLockedBalance failed, zero balance";
+        return tecUNFUNDED_PAYMENT;
+    }
+
+    STAmount priorLockedBalance{sfLockedBalance, deltaAmt.issue()};
+    if (sleLine->isFieldPresent(sfLockedBalance))
+        priorLockedBalance =
+            high ? -(*sleLine)[sfLockedBalance] : (*sleLine)[sfLockedBalance];
+
+    uint32_t priorLockCount = 0;
+    if (sleLine->isFieldPresent(sfLockCount))
+        priorLockCount = sleLine->getFieldU32(sfLockCount);
+
+    uint32_t finalLockCount = priorLockCount + deltaLockCount;
+    STAmount finalLockedBalance = priorLockedBalance + deltaAmt;
+
+    if (finalLockedBalance > balance)
+    {
+        JLOG(j.trace()) << "trustAdjustLockedBalance: "
+                        << "lockedBalance(" << finalLockedBalance
+                        << ") > balance(" << balance << ") = true\n";
+        return tecINSUFFICIENT_FUNDS;
+    }
+
+    if (finalLockedBalance < beast::zero)
+        return tecINTERNAL;
+
+    // check if there is significant precision loss
+    if (!isAddable(balance, deltaAmt) ||
+        !isAddable(priorLockedBalance, deltaAmt) ||
+        !isAddable(finalLockedBalance, balance))
+        return tecPRECISION_LOSS;
+
+    // sanity check possible overflows on the lock counter
+    if ((deltaLockCount > 0 && priorLockCount > finalLockCount) ||
+        (deltaLockCount < 0 && priorLockCount < finalLockCount) ||
+        (deltaLockCount == 0 && priorLockCount != finalLockCount))
+        return tecOVERSIZE;
+
+    // we won't update any SLEs if it is a dry run
+    if (dryRun)
+        return tesSUCCESS;
+
+    if constexpr (
+        std::is_same<V, ApplyView>::value &&
+        std::is_same<S, std::shared_ptr<SLE>>::value)
+    {
+        if (finalLockedBalance == beast::zero || finalLockCount == 0)
+        {
+            sleLine->makeFieldAbsent(sfLockedBalance);
+            sleLine->makeFieldAbsent(sfLockCount);
+        }
+        else
+        {
+            sleLine->setFieldAmount(
+                sfLockedBalance,
+                high ? -finalLockedBalance : finalLockedBalance);
+            sleLine->setFieldU32(sfLockCount, finalLockCount);
+        }
+
+        view.update(sleLine);
+    }
+
+    return tesSUCCESS;
+}
+
+/** Check if a set of accounts can freely exchange the specified token.
+    Read only, does not change any ledger object.
+    May be called with ApplyView or ReadView.
+    (including unlocking) is forbidden by any flag or condition.
+    If parties contains 1 entry then noRipple is not a bar to xfer.
+    If parties contains more than 1 entry then any party with noRipple
+    on issuer side is a bar to xfer.
+*/
+template <class V>
+[[nodiscard]] TER
+trustTransferAllowed(
+    V& view,
+    std::vector<AccountID> const& parties,
```

Comment:
why making this an array when `parties` has at most a size of 2?

if that's the case, should we assert that `parties.size() <= 2 `at the beginning of the function?